Bat scripts to work with JAVA_HOME with parantheses

pgomulka opened the pull request:
the elasticsearch.bat and elasticsearch-env.bat won't work if JAVA contains parentheses. This seems to be the limitation of FOR /F IN (command) DO syntax.
The JAVA variable present in a command contains a path to a binary to start elasticsearch (with spaces & parans). We can workaround the problem of spaces and parentheses in this path by referring this variable with a CALL command.
Note that executing binaries with CALL is an undocumented behaviour (but works)

closes #38578
closes #38624
closes #33405
closes #30606

pgomulka committed:
Bat scripts to work with JAVA_HOME with parantheses

the elasticsearch.bat and elasticsearch-env.bat won't work if JAVA
contains parentheses. This seems to be the limitation of FOR /F IN
(command) DO syntax.
The JAVA variable present in a command contains a path to a binary to
start elasticsearch (with spaces & parans). We can workaround the
problem of spaces and parentheses in this path by referring this
variable with a CALL command.
Note that executing binaries with CALL is an undocumented behaviour (but
works)

elasticmachine commented:
Pinging @elastic/es-core-infra

pgomulka commented:
ok to test

pgomulka committed:
Merge remote-tracking branch 'upstream/master' into feature/windows_startup_scripts

rjernst reviewed (commented):
Sorry this review never got published before. Some of these comments might have already been addressed.

pgomulka commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -193,6 +192,39 @@ public void test50StartAndStop() throws IOException {
+                Archives.runElasticsearch(installation, sh);
+
+                Archives.stopElasticsearch(installation);
+            } catch (IOException e) {
```
it is because this code is inside lambda` onWindows( ()-> {here} )`

rjernst replied:
We should change the signature of `PlatformAction.run` to have a checked `Exception`, so that we don't need these intermediate try/catches.

pgomulka commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -193,6 +192,39 @@ public void test50StartAndStop() throws IOException {
+                                             .filter(path -> path.contains("Java") == false)
+                                             .collect(joining(";"));
+
+                sh.runIgnoreExitCode("cmd /c mklink /D 'C:\\Program Files (x86)\\java' $Env:JAVA_HOME");
```
you mean the symbolic link creation? Is the PS 5 available since certain windows version?

rjernst replied:
Why do we ingnore the exit code? If this fails, the test cannot proceed right?

pgomulka replied:
good spot, test should not proceed if this fails

pgomulka committed:
refactor run method

pgomulka commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -193,6 +192,39 @@ public void test50StartAndStop() throws IOException {
+            final String javaHome = sh.run("$Env:JAVA_HOME").stdout.trim();
+
+            try {
+                final String newPath = Arrays.stream(originalPath.split(";"))
```
so the vagrant images are guaranteed not to have java inside PATH? I feel like it wouldn't harm to make sure PATH does not have java in there.

rjernst replied:
No, java is in the PATH. I don't think we should change that. In 6.7 I made sure all the tests pass in JAVA_HOME, to avoid triggering the deprecation warning output starting in that version when using java on the PATH. In 7.0+ we no longer look at PATH at all to find java.

pgomulka committed:
Linux version of java with spaces

pgomulka committed:
compile fix

pgomulka committed:
fix temp path

pgomulka committed:
Merge branch 'master' into feature/windows_startup_scripts

pgomulka committed:
minor cleanup

pgomulka committed:
not looking into PATH at all

pgomulka committed:
compile error

pgomulka committed:
compile errors

pgomulka commented:
@elasticmachine run elasticsearch-ci/packaging-sample

pgomulka committed:
fix the linux version of the test

pgomulka committed:
Merge branch 'master' into feature/windows_startup_scripts

pgomulka committed:
fix linux test

pgomulka committed:
fixing failing tests

pgomulka committed:
permissions for tmp file

pgomulka committed:
chown temp file

pgomulka committed:
Merge remote-tracking branch 'upstream/master' into feature/windows_startup_scripts

pgomulka committed:
compile error

pgomulka commented:
@rjernst would you be able to review this again?

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+            } finally {
+                //clean up sym link
+                sh.runIgnoreExitCode("cmd /c del /F /Q 'C:\\Program Files (x86)\\java' ");
+                sh.getEnv().put("JAVA_HOME", javaHome);
```
No need for setting JAVA_HOME here.

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+
+                FileUtils.rm(Paths.get(temp));
+            } finally {
+                sh.getEnv().put("JAVA_HOME", javaHome);
```
ditto, this can be removed

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+                String pluginListCommand = installation.bin + "/elasticsearch-plugin list";
+                Result result = sh.run(pluginListCommand);
+                assertThat(result.exitCode, equalTo(0));
+
```
nit: extra blank line

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+
+                // Create temporary directory with a space and link to java binary.
+                // Use it as java_home
+                final String temp = sh.runIgnoreExitCode("mktemp -d --suffix=\"java home\"").stdout.trim();
```
Create a temp directory seems error prone, since it may have noexec on some systems. I think this test can use any arbitrary path since it will clean it up afterwards?

pgomulka replied:
we create temp directories when starting up as well. If some systems are not supporting this, it would be better to catch that during testing.
Do you have any specific systems we support in mind that would fail on this?

rjernst replied:
Yes we do create temp directories, but we don't try executing anything from them. Why we do we need this to be complicated creating a special temp dir when we can just use an arbitrary path and clean it up afterwards?

pgomulka replied:
I see, we can use a regular directory for this. Should be easy. I was thinking that `mktemp` is ok as we use it through a lot of `bats` tests at the moment. Will keep an eye on further migration to avoid it.

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+                // Create temporary directory with a space and link to java binary.
+                // Use it as java_home
+                final String temp = sh.runIgnoreExitCode("mktemp -d --suffix=\"java home\"").stdout.trim();
+                final String java = sh.runIgnoreExitCode("which java").stdout.trim();
```
Look for $SYSTEM_JAVA_HOME. See other tests that set JAVA_HOME. We no longer use java from the path.

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+        Platforms.onWindows(() -> {
+            final Shell sh = new Shell();
+            final String originalPath = sh.run("$Env:PATH").stdout.trim();
+            final String javaHome = sh.run("$Env:JAVA_HOME").stdout.trim();
```
JAVA_HOME does not always exist. You need to look for SYSTEM_JAVA_HOME

pgomulka replied:
will remove that at all. As per your comment [here](https://github.com/elastic/elasticsearch/pull/39712/files#r270126523) we don't need to clean it up

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +194,70 @@ public void test52BundledJdkRemoved() throws IOException {
+
+        Platforms.onWindows(() -> {
+            final Shell sh = new Shell();
+            final String originalPath = sh.run("$Env:PATH").stdout.trim();
```
No need to mess with the PATH

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Archives.java:
```diff
@@ -255,7 +253,11 @@ private static void verifyDefaultInstallation(Installation es, Distribution dist
 
-    public static void runElasticsearch(Installation installation, Shell sh) throws IOException {
+    public static void runElasticsearch(Installation installation) throws Exception {
+        runElasticsearch(installation, new Shell());
```
I just removed this recently. We need to always pass in a Shell, so we can control when/if JAVA_HOME is set. Please revert this part.

rjernst reviewed (changes requested):
There are a lot of unrelated changes to method signatures and whitespace, Can you clean those up so this PR represents just he actual changes to JAVA_HOME with parentheses?

pgomulka committed:
code review follow up

pgomulka commented:
@rjernst I have removed the unnecessary whitespace changes. We have however already agreed here https://github.com/elastic/elasticsearch/pull/39712#discussion_r265373126 to throw Exception from `PlatformAction`. This would help avoiding unnecessary try/catch blocks.
Also addressed the code review comments

rjernst commented:
Sure, the signature changes are fine, I see the change now.

pgomulka committed:
remove random tmp creation

pgomulka committed:
refactor tmp creation to regular direcotry

pgomulka committed:
fix cleanup

rjernst commented on qa/vagrant/src/main/java/org/elasticsearch/packaging/test/ArchiveTestCase.java:
```diff
@@ -192,7 +196,63 @@ public void test52BundledJdkRemoved() throws IOException {
+
+            } finally {
+                //clean up sym link
+                sh.runIgnoreExitCode("cmd /c del /F /Q 'C:\\Program Files (x86)\\java' ");
```
I don't think we want to ignore the exit? If this fails, later tests could fail?

pgomulka replied:
Actuality I thought that if this fails the rest of the test should continue. JAVA_HOME should not propagate so it won't be visible if failed

rjernst replied:
If it fails, something is messed up. I think we should stop testing.

rjernst reviewed (approved):
LGTM one last request

pgomulka committed:
don't ignore cleanup failure

pgomulka commented:
This needs to be backported to 6.7 branch (for 6.7.2 version) and 7.x (for 7.1)

jpountz commented:
@pgomulka  I'm assuming there is nothing left to backport and removed the backport pending label.
